Implement `useFills` hook

michalczaplinski opened the pull request:
**What**:

Implement the `useFills` React hook which is a prerequisite for creating `<Slot/>` components for a Slot and Fill pattern. This will allow theme creators to create "slots" where the theme users can insert their "fill" components. The "fill" can be any React component.

**Why**:

To allow users to create "slots" in themes that can then be filled by "fills". More rationale[ in the Feature Discussion](https://community.frontity.org/t/slot-and-fill/895)

**How**:

Following the Implementation Proposal: https://community.frontity.org/t/slot-and-fill/895/8

**Tasks**:

- [x] Code
- [x] TypeScript
- [x] Changeset
- [x] Community discussions
- [x] TSDocs
- [x] Unit tests
- [x] Move to "frontity" package
- [x] Add namespaces to `state.fills` and `libraries.fills`
- [x] End to end tests
- [x] TypeDocs
- [x] Documentation https://github.com/frontity/gitbook-docs/pull/75 (still WIP)

Unrelated:
- ~~TypeScript tests~~
- ~~Update starter themes~~
- ~~Update other packages~~


<!-- Feel free to add any additional comments. -->

michalczaplinski committed:
hooks: use-fills first steps

michalczaplinski commented on packages/hooks/use-fills.tsx:
```diff
@@ -0,0 +1,39 @@
+}
+
+const useFills = (name) => {
+  const { state, libraries } = useConnect<Fills>();
```
@orballo Could you tell me if this is the right approach to typing the output of `useConnect` ?

Or should I import the types from somewhere ?

orballo replied:
That's the right way

luisherranz commented on packages/hooks/package.json:
```diff
@@ -22,5 +22,9 @@
+  },
+  "devDependencies": {
+    "react-test-renderer": "^16.13.1",
+    "@frontity/connect": "1.0.5"
```
Please don't add fixed dependencies.

luisherranz commented on packages/hooks/use-fills.tsx:
```diff
@@ -0,0 +1,39 @@
+
+  Object.values(state.fills)
+    // Match only the fills for this name.
+    .find((fill) => fill.name === name)
```
This is `fill.slot`. `fill.name` doesn't exist. I have fixed it in the IP.

luisherranz commented on packages/hooks/use-fills.test.tsx:
```diff
@@ -0,0 +1,48 @@
+describe("test", () => {
+  it("should work", () => {
+    const Comp = () => {
+      const fills = useFills("test fill");
```
This is "slot 1". We want all fills for "slot 1" beucase this component is "Slot 1".

luisherranz commented on packages/hooks/use-fills.test.tsx:
```diff
@@ -0,0 +1,48 @@
+    const Comp = () => {
+      const fills = useFills("test fill");
+
+      return fills.map(({ Fill, props }) => <Fill key={props.id} {...props} />);
```
`props.id` is not guaranteed to exist. Let's add the fill's name as key: https://community.frontity.org/t/slot-and-fill/895/10?u=luisherranz

michalczaplinski committed:
hooks: Add more tests cases and update the implementation of useFill()

michalczaplinski committed:
hooks: Refactor map + filter to .reduce

michalczaplinski committed:
hooks: Add better TS declaration for useFills

luisherranz commented on packages/hooks/use-fills.tsx:
```diff
@@ -1,39 +1,73 @@
+  }
+
+  if (!Object.values(state.fills).find((slot) => slot.slot === name)) {
+    warn(`Could not find any slot with name ${name}. Is this what you want?`);
```
This warning is not useful because the theme will declare many slots and most of them will be empty.

michalczaplinski committed:
hooks: useFill - Add more tests

michalczaplinski committed:
hooks: useFill - test priority

michalczaplinski commented:
As this is nearing finish, there are still some remaining questions:

1. Do we want to implement the [data assertions](https://community.frontity.org/t/slot-and-fill/895/3?u=mmczaplinski) here as well?
2. Is the current behaviour or returning the `[]` as a "default fill" okay?
3. Are we deprecating exporting the `roots` property of the state and does that mean refactoring e.g. the GTM package?

michalczaplinski committed:
hooks: useFill - Add type declarations

michalczaplinski commented on packages/hooks/package.json:
```diff
@@ -18,9 +18,14 @@
     "test:ci": "../../node_modules/.bin/jest --passWithNoTests",
     "test": "../../node_modules/.bin/jest --watch",
-    "test:inspect": "node --inspect ./node_modules/jest/bin/jest.js --watch --no-cache --runInBand"
+    "test:inspect": "node --inspect ../../node_modules/jest/bin/jest.js --watch --no-cache --runInBand"
```
This was incorrect before for some reason.

michalczaplinski commented on packages/types/src/package.ts:
```diff
@@ -15,6 +24,7 @@ export interface Package {
       packages?: string[];
       platform?: "client" | "server";
       rendering?: "ssr" | "csr";
+      fills: { [key: string]: Fill };
```
I'm not entirely sure if this is how I'm supposed to add the types here. @DAreRodz Do you have some experience with this?

luisherranz replied:
You put `fills` inside `state.frontity`. This is wrong, it's not `state.frontity.fills`, it is `state.fills`.

michalczaplinski replied:
Right, my bad, that's true. 

But assuming that I put them in `state.fills` my question was more about the general idea. Is this the right place to put the types for `fills`? I think so, but would just like to confirm.

luisherranz replied:
Yes, I think the placement and the types are right 👍

luisherranz commented:
> 1. Do we want to implement the (data assertions)[https://community.frontity.org/t/slot-and-fill/895/3?u=mmczaplinski] here as well?

No, that's not included in the Implementation Proposal: https://community.frontity.org/t/slot-and-fill/895/6

> 2. Is the current behaviour or returning the `[]` as a "default fill" okay?

Yes, if there are no fills for that slot name an empty array is fine so people can still use `fills.map`.

> 3. Are we deprecating exporting the `roots` property of the state and does that mean refactoring e.g. the GTM package?

I suggested to rename it (while keeping backward compatibility), but that's outside of the scope of the Implementation Proposal.

michalczaplinski commented:
> No, that's not included in the Implementation Proposal: https://community.frontity.org/t/slot-and-fill/895/6

Right, I understand, but there were several updates to the implementation proposal and as the data assertions were mentioned at some point it was not 100% clear that we are not implementing them at this point 🙂

michalczaplinski committed:
hooks: useFills - add tests for removing the cases from the state

michalczaplinski committed:
hooks: useFills - fix types in packages

michalczaplinski committed:
hooks: useFills - fix the type definitions

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into use-fill

michalczaplinski committed:
hooks: fix the comment

michalczaplinski commented:
Really not sure why pushing with `--force`  closed the PR 🤷

michalczaplinski commented:
The diff is showing the changes from the base branch (`use-connect`) relative to `dev` even though I'm trying to merge `use-connect` it. I'm not sure how to unfuck this.

Let's wait until the `use-connect` branch is merged into `dev`, so then we can change the base branch to `dev` and the diff should be clean.

michalczaplinski committed:
Merge branch 'dev' into use-fill

michalczaplinski committed:
Add changeset

michalczaplinski commented:
FYI, The [docs](https://github.com/frontity/gitbook-docs/pull/69) are not 100% finished for this but the code can be reviewed.

luisherranz commented on packages/hooks/use-fills.tsx:
```diff
@@ -0,0 +1,71 @@
+      }))
+
+      // Match only the fills for this name.
+      .filter(({ slot }) => slot === name)
```
You can do this first.

luisherranz commented on packages/hooks/use-fills.tsx:
```diff
@@ -0,0 +1,71 @@
+  };
+}
+
+const useFills = (name: string) => {
```
This is missing TSDocs.

Maybe explicitly define the return type as well, so we cannot break it in a refactoring.

michalczaplinski replied:
👍   on the TSDoc.

The return type is:
<img width="381" alt="Screenshot 2020-05-22 at 14 42 44" src="https://user-images.githubusercontent.com/5417266/82704116-80227700-9c3a-11ea-8f0b-aa19a42f7d43.png">

because I assign a type here: https://github.com/frontity/frontity/blob/e2056a31a608e05d58f60c7ec98ee4ec5934209e/packages/hooks/use-fills.tsx#L47-L49

I need to specify the type of that accumulator so that I get type checks inside of the `.reduce` callback.

So the type inference figures out the correct return type of `useFills` for me. Is there a problem with this approach that I'm not aware of?

luisherranz replied:
Explicit return types make impossible to break the function (type-wise) in a subsequent refactoring.

luisherranz commented on packages/hooks/use-fills.tsx:
```diff
@@ -0,0 +1,71 @@
+import { useConnect, warn } from "frontity";
+import { Fill } from "@frontity/types";
+
+interface FillsPackage {
```
Instead of redeclaring the type, you can import `Package`. That way if we change the origin type, it changes here as well.

michalczaplinski replied:
👍

luisherranz commented on packages/types/src/package.ts:
```diff
@@ -1,3 +1,12 @@
+export interface Fill {
+  slot: string;
```
Can you add TSDocs to this file as well?

michalczaplinski replied:
I've tried to do this, but don't really see what benefit it brings to add the TSDoc here, to be honest 🤔 

TSDoc does not really have a way to "document" the interfaces. I also don't really know what I should document here? It's just a type for the `Fill`, it should be self-evident what it is.

Could you elaborate on what you had in mind?

luisherranz replied:
Let's not go over this, please. We need to add TSDocs to all our interfaces, functions, classes, types...

Interfaces are documented in the same way than classes.

luisherranz commented on packages/hooks/use-fills.test.tsx:
```diff
@@ -0,0 +1,321 @@
+
+    expect(app.toJSON()).toMatchSnapshot();
+  });
+});
```
Great tests.

The only one I miss is to check if it doesn't include a fill that doesn't match the slot name.

luisherranz reviewed (changes requested):
Awesome work.

I've added some details.

And as it seems that we are going to start using this package a lot, I would like to take this opportunity to mark this as side effect free.

https://webpack.js.org/guides/tree-shaking/

Could you please do so (`sideEffects: false`) and create and index file with all the exports?
Thanks!

luisherranz commented on .changeset/lucky-rocks-fetch.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@frontity/hooks": minor
```
Totally my fault I didn't mention this earlier, but could you please move it to the main `"frontity"` package?

michalczaplinski committed:
hooks: useFills - reorder state.fills methods

michalczaplinski committed:
hooks: useFills - mark the module as side-effect-free

michalczaplinski committed:
hooks: useFills - Remove unnecessary type redeclaration

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into use-fill

michalczaplinski committed:
hooks: useFills - add TSDoc

luisherranz commented on packages/hooks/use-fills.tsx:
```diff
@@ -1,6 +1,13 @@
 import { useConnect, warn } from "frontity";
 import { Fill, Package } from "@frontity/types";
```
Don't use `"@frontity/types"`, use `"frontity/types"`.

Commit:
Remove old fills type

Commit:
Add TSDocs and explicit return type

luisherranz reviewed (dismissed):
I've done the TSDocs myself so you can get the idea of how they work. I've also added the explicit return type so you can see what I mean.

I think the only thing missing is to move this to the main `"frontity"` package and the unit test.

luisherranz commented:
I'm going to take over and finish this because I also want to add namespaces to both `state.fills` and `libraries.fills` as I have explained in the [feature discussion](https://community.frontity.org/t/slot-and-fill/895/17?u=luisherranz).

luisherranz committed:
Fix hooks version in e2e project

luisherranz committed:
Fix dependency mismatch

luisherranz committed:
Remove publish command from lerna

We don't use it anymore, we use changets now.

luisherranz committed:
Update TypeScript version

luisherranz committed:
Fix ResolveActions in TypeScript 3.9

luisherranz committed:
Merge branch 'dev' into use-fill

luisherranz committed:
Move useFills to frontity package

luisherranz committed:
Add TSDocs to Package types

luisherranz committed:
Add more TSDocs and update some types

luisherranz committed:
Switch to using namespaces in fills [WIP]

DAreRodz commented:
@luisherranz, I've assigned you the issue related to this PR (#431) and moved it to **in progress**.

luisherranz commented:
Thanks David! I totally have to get used to the board 😅

luisherranz committed:
Finish switching using to namespaces in fills

luisherranz committed:
Rename to slot-and-fill

luisherranz committed:
Remove store.fills from app initialization

luisherranz committed:
Add e2e package

luisherranz committed:
Update Cypress and Snapshots plugin

luisherranz committed:
Add e2e tests for useFills

luisherranz committed:
Merge branch 'dev' into use-fill

luisherranz committed:
Update changesets

luisherranz committed:
Add package-lock again

I deleted it by mistake.

luisherranz committed:
Avoid usage of flatMap

It's not supported on Node 8 and 10 and I prefer not to use babel for jest.

luisherranz committed:
Fix TypeScript version

luisherranz committed:
Add type to reduce

luisherranz commented:
@michalczaplinski, this is ready for code review. I cannot add you as reviewer but could you please check it out?

I saw that TS 3.9 included some breaking changes so I did a small refactor of the types. Then I saw that it also broke our `source` data types, so I removed the upgrade and fixed the version to TS 3.8 to open that in another PR.

Other than that I moved the hook to `"frontity"`, refactored to use namespaces inside `state.fills` and `libraries.fills`, added more TSDocs and some simple e2e tests.

I still want to:

- Explore how to export the markdown from TSDocs.
- Take a look at the Doc PR: https://github.com/frontity/gitbook-docs/pull/75

But the code should be fine now.

luisherranz committed:
Fix missing default

michalczaplinski commented:
Just FYI:

> I still want to:
> - Take a look at the Doc PR: frontity/gitbook-docs#75

The docs are mostly there, but I think that they still miss some more details. Are you planning to you finish those?

michalczaplinski commented on lerna.json:
```diff
@@ -1,12 +1,4 @@
-      "githubRelease": true,
-      "message": "chore(release): publish [skip ci]"
-    }
-  }
```
I'm not really familiar with publishing to npm with lerna, but is there some change to how we're going to do that from now on?

luisherranz replied:
We are using `changesets` instead of `lerna` to publish our packages so that is not necessary anymore.

Sorry to have done that here but something failed in lerna, I don't remember exactly what, so I removed that code to fix it and because we don't use it anymore.

michalczaplinski replied:
oh I see, thanks :)

michalczaplinski commented on packages/connect/src/index.d.ts:
```diff
@@ -9,9 +9,15 @@ interface ObserveOptions {
   lazy?: boolean;
 }
 
+type Function = (...args: any) => any;
```
There already exists an interface `Function` in TS which describes a generic function. Do we need this?

luisherranz replied:
Oh, thanks. I'll remove that.

michalczaplinski committed:
useFill: fix typos in comments

michalczaplinski commented on packages/frontity/src/utils/slot-and-fill/use-fills.tsx:
```diff
@@ -0,0 +1,75 @@
+    // Flat all the fills and turn them into entries.
+    Object.values(state.fills || {})
+      .reduce<[string, StateFill][]>(
+        (fills, fill) => [...fills, ...Object.entries(fill)],
```
Just a nit but wouldn't this be clearer as:
```
 (fills, namespace) => [...fills, ...Object.entries(namespace)],
```

luisherranz replied:
Oh, absolutely. I'll change it.

michalczaplinski committed:
types: Update comment wording

michalczaplinski commented on packages/core/src/app/index.tsx:
```diff
@@ -15,9 +15,6 @@ const App: React.FunctionComponent<Props> = ({ store, helmetContext = {} }) => {
         ))}
-        {Object.entries(store.fills).map(([namespace, Fill]) => (
-          <Fill key={namespace} />
-        ))}
```
So, the `Fill` components were not supposed to be placed there in the first place?

luisherranz replied:
They were. It was the first approach to integrate Slot and Fill and the way we used it in the old framework.

But then I decided to use `state.fills` and `libraries.fills` instead, which makes much more sense for Frontity.

It is explained in https://community.frontity.org/t/slot-and-fill/895

michalczaplinski reviewed (commented):
I can't "approve" or request changes  because I'm the author, but I have a couple of comments.

luisherranz commented:
> Are you planning to you finish those?

Yes. But I want to do this first to export the markdown of the TSDocs and see how we can use it: https://community.frontity.org/t/code-documentation-workflow/1633/9

michalczaplinski commented:
By the way, just to clarify, there were just a couple of comments but it LGTM 👍

michalczaplinski commented:
Oh, sorry, I've just realized that I think you forgot to include the changesets for all the other packages that you updated 🙂

luisherranz committed:
Remove unnecessary Function type declaration

luisherranz committed:
Rename fill to namespace in reduce callback

luisherranz committed:
Rename wrong package on changeset

luisherranz committed:
Revert tsconfig target update

Because Node doesn't support some of the 2019 updates and we are
not transpiling the code on Jest.

luisherranz committed:
Add new changesets

luisherranz committed:
Merge branch 'dev' into use-fill

luisherranz commented:
> Oh, sorry, I've just realized that I think you forgot to include the changesets for all the other packages that you updated 🙂

Thanks for the heads up! I have updated/added proper changesets 👍

luisherranz committed:
Update package-lock

luisherranz commented:
As the [Typedocs approach hasn't worked as we thought](https://community.frontity.org/t/adopting-a-standard-way-to-document-frontity-apis/1662/11?u=luisherranz), I'm going to continue with this PR.

luisherranz committed:
Merge branch 'dev' into use-fill

luisherranz committed:
Add namespaces to React keys

Because if not, two fills that have the same name but belong to two
different namespaces could collide.

luisherranz committed:
Add debug mode

luisherranz committed:
Add support for debug mode

luisherranz committed:
Update `debug` in snapshot

luisherranz commented:
I have reused [your styles](https://github.com/wp-pwa/saturn-theme/blob/36852a9d8d39302e1603b15b879e16b0f38f80a4/src/shared/components/SlotInjector/index.js#L89-L93) @darerodz 🙂

I have added:

- Support for debug mode using `state.frontity.debug = true`.
- Unique `key`s by using the namespace as well as the fill name to avoid problems in case two packages using different namespaces use the same fill name.

This should be ready to merge. @michalczaplinski feel free to review these last two improvements again if you want.

I'll finish the docs tomorrow.

luisherranz committed:
Fix frontity mock

luisherranz committed:
Merge branch 'dev' into use-fill
